cedar-policy-core/src/parser/cst_to_ast.rs

```diff
@@ -147,103 +155,80 @@ impl Node<Option<cst::Policy>> {
-        //     we also throw in this error. Ideally we'd do this if the partially parsed AST included
-        //     any template slots at all, but we don't have an easy mechanism for that currently
-
-        // case 2 first: if we failed to parse, but the parse errors include a
-        // `SlotsInConditionClause`, we can report that as `UnexpectedTemplate`
-        let new_errs = errs
-            .iter()
-            .filter_map(|err| match err {
-                ParseError::ToAST(err) => match err.kind() {
-                    ToASTErrorKind::SlotsInConditionClause { slot, .. } => Some(ToASTError::new(
-                        ToASTErrorKind::UnexpectedTemplate { slot: slot.clone() },
-                        err.source_loc().clone(),
-                    )),
-                    _ => None,
-                },
-                _ => None,
-            })
-            .collect::<Vec<_>>();
-        errs.extend(new_errs);
-
-        // now case 1: if `policy` is `Some(Err(e))`, there is a slot where there shouldn't be;
-        // report that as `UnexpectedTemplate`
-        match policy {
-            Some(Err(ast::UnexpectedSlotError::FoundSlot(slot))) => {
-                errs.push(ToASTError::new(
+    pub fn to_policy(&self, id: ast::PolicyID) -> Result<ast::StaticPolicy, ParseErrors> {
+        let maybe_template = self.to_policy_template(id);
+        let maybe_policy = maybe_template.map(ast::StaticPolicy::try_from);
+        match maybe_policy {
+            // Successfully parsed a static policy
+            Ok(Ok(p)) => Ok(p),
+            // The source parsed as a template, but not a static policy
+            Ok(Err(ast::UnexpectedSlotError::FoundSlot(slot))) => {
+                Err(ParseErrors::singleton(ToASTError::new(
                     ToASTErrorKind::UnexpectedTemplate {
                         slot: slot.id.into(),
                     },
                     slot.loc.unwrap_or_else(|| self.loc.clone()),
-                ));
-                None
+                )))
+            }
+            // The source failed to parse completely. If the parse errors include
+            // `SlotsInConditionClause` also add an `UnexpectedTemplate` error.
+            Err(mut errs) => {
+                let new_errs = errs
+                    .iter()
+                    .filter_map(|err| match err {
+                        ParseError::ToAST(err) => match err.kind() {
+                            ToASTErrorKind::SlotsInConditionClause { slot, .. } => {
+                                Some(ToASTError::new(
+                                    ToASTErrorKind::UnexpectedTemplate { slot: slot.clone() },
+                                    err.source_loc().clone(),
+                                ))
+                            }
+                            _ => None,
+                        },
+                        _ => None,
+                    })
+                    .collect::<Vec<_>>();
```

Comment:
I think we can avoid this `.collect()`, and just pass an iterator straight to `.extend()`

Reply:
Based on the compiler error message, it seems this is required since the input to `extend` is an iterator over `errs` (which is what's being extended).